File: mcp/ts/src/tools/ingest-batch.test.ts
// SPDX-License-Identifier: Apache-2.0

import { describe, expect, it } from 'vitest'
import type { MemoryClient, ProgressEmitter } from '../memory-client.js'
import { ingestBatchTool } from './ingest-batch.js'
import type { ToolContext } from './types.js'

const noopClient = (): MemoryClient => ({
  mode: 'local',
  remember: async () => ({}),
  recall: async () => ({}),
  search: async () => ({}),
  ask: async () => ({}),
  ingestFile: async () => ({}),
  ingestUrl: async () => ({}),
  extract: async () => ({}),
  reflect: async () => ({}),
  consolidate: async () => ({}),
  createBrain: async () => ({}),
  listBrains: async () => ({}),
  close: async () => undefined,
})

type IngestFileCall = {
  readonly path: string
  readonly brain: string | undefined
  readonly as: string | undefined
}

describe('memory_ingest_batch tool', () => {
  it('processes a batch of 3 valid files and returns all succeeded', async () => {
    const calls: IngestFileCall[] = []
    const client = noopClient()
    client.ingestFile = async (args: { path: string; brain?: string; as?: string }) => {
      calls.push({ path: args.path, brain: args.brain, as: args.as })
      return {
        status: 'completed',
        document_id: `doc-${args.path}`,
        hash: `hash-${args.path}`,
        byte_size: 100,
      }
    }

    const result = await ingestBatchTool.handler(
      {
        files: [
          { path: '/tmp/a.md' },
          { path: '/tmp/b.md' },
          { path: '/tmp/c.md' },
        ],
      },
      client,
      {},
    )

    const payload = result.structuredContent as {
      total: number
      succeeded: number
      failed: number
      results: { path: string; status: string; documentId: string }[]
    }
    expect(payload.total).toBe(3)
    expect(payload.succeeded).toBe(3)
    expect(payload.failed).toBe(0)
    expect(payload.results).toHaveLength(3)
    expect(calls).toHaveLength(3)
    expect(payload.results[0]?.status).toBe('success')
    expect(payload.results[0]?.documentId).toBe('doc-/tmp/a.md')
  })

  it('isolates per-file errors: file 2 fails but file 1 and 3 succeed', async () => {
    const client = noopClient()
    let callIndex = 0
    client.ingestFile = async (args: { path: string }) => {
      callIndex++
      if (callIndex === 2) {
        throw new Error('file not found')
      }
      return {
        status: 'completed',
        document_id: `doc-${args.path}`,
        hash: `hash-${args.path}`,
        byte_size: 50,
      }
    }

    const result = await ingestBatchTool.handler(
      {
        files: [
          { path: '/tmp/a.md' },
          { path: '/tmp/missing.md' },
          { path: '/tmp/c.md' },
        ],
      },
      client,
      {},
    )

    const payload = result.structuredContent as {
      total: number
      succeeded: number
      failed: number
      results: { path: string; status: string; error?: string }[]
    }
    expect(payload.total).toBe(3)
    expect(payload.succeeded).toBe(2)
    expect(payload.failed).toBe(1)
    expect(payload.results[0]?.status).toBe('success')
    expect(payload.results[1]?.status).toBe('error')
    expect(payload.results[1]?.error).toBe('file not found')
    expect(payload.results[2]?.status).toBe('success')
  })

  it('emits progress notifications for each file processed', async () => {
    const client = noopClient()
    client.ingestFile = async () => ({
      status: 'completed',
      document_id: 'doc',
      hash: 'hash',
      byte_size: 10,
    })

    const progressCalls: { progress: number; message?: string }[] = []
    const ctx: ToolContext = {
      progress: (p: number, m?: string) => {
        progressCalls.push({ progress: p, message: m })
      },
    }

    await ingestBatchTool.handler(
      {
        files: [
          { path: '/a.md' },
          { path: '/b.md' },
          { path: '/c.md' },
        ],
      },
      client,
      ctx,
    )

    expect(progressCalls).toHaveLength(3)
    expect(progressCalls[0]?.progress).toBe(1)
    expect(progressCalls[0]?.message).toBe('1/3 /a.md')
    expect(progressCalls[1]?.progress).toBe(2)
    expect(progressCalls[1]?.message).toBe('2/3 /b.md')
    expect(progressCalls[2]?.progress).toBe(3)
    expect(progressCalls[2]?.message).toBe('3/3 /c.md')
  })

  it('applies brain parameter to all files in the batch', async () => {
    const brainsSeen: (string | undefined)[] = []
    const client = noopClient()
    client.ingestFile = async (args: { brain?: string }) => {
      brainsSeen.push(args.brain)
      return { status: 'completed', document_id: 'doc', hash: 'h', byte_size: 1 }
    }

    await ingestBatchTool.handler(
      {
        files: [{ path: '/a.md' }, { path: '/b.md' }],
        brain: 'test-brain',
      },
      client,
      {},
    )

    expect(brainsSeen).toEqual(['test-brain', 'test-brain'])
  })

  it('handles duplicate file paths: both entries are processed independently', async () => {
    const calls: string[] = []
    const client = noopClient()
    client.ingestFile = async (args: { path: string }) => {
      calls.push(args.path)
      return {
        status: 'completed',
        document_id: `doc-${calls.length}`,
        hash: `hash-${calls.length}`,
        byte_size: 42,
      }
    }

    const result = await ingestBatchTool.handler(
      {
        files: [
          { path: '/data/same.md' },
          { path: '/data/same.md' },
        ],
      },
      client,
      {},
    )

    const payload = result.structuredContent as {
      total: number
      succeeded: number
      failed: number
      results: { path: string; status: string }[]
    }
    expect(payload.total).toBe(2)
    expect(payload.succeeded).toBe(2)
    expect(payload.failed).toBe(0)
    expect(calls).toHaveLength(2)
    expect(calls[0]).toBe('/data/same.md')
    expect(calls[1]).toBe('/data/same.md')
  })

  it('rejects empty files array via zod validation', () => {
    const parseResult = ingestBatchTool.inputSchema.safeParse({ files: [] })
    expect(parseResult.success).toBe(false)
  })

  it('rejects files array exceeding 50 entries via zod validation', () => {
    const files = Array.from({ length: 51 }, (_, i) => ({ path: `/file-${i}.md` }))
    const parseResult = ingestBatchTool.inputSchema.safeParse({ files })
    expect(parseResult.success).toBe(false)
  })

  it('accepts exactly 50 files via zod validation', () => {
    const files = Array.from({ length: 50 }, (_, i) => ({ path: `/file-${i}.md` }))
    const parseResult = ingestBatchTool.inputSchema.safeParse({ files })
    expect(parseResult.success).toBe(true)
  })
})

File: go/cmd/memory-mcp/tools/ingest_batch_test.go
// SPDX-License-Identifier: Apache-2.0

package tools

import (
	"context"
	"encoding/json"
	"fmt"
	"sync"
	"testing"
	"time"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

// mockMemoryClient is a minimal MemoryClient for testing batch ingest.
type mockMemoryClient struct {
	ingestFileFn func(ctx context.Context, args IngestFileArgs, progress ProgressEmitter) (*IngestResult, error)
}

func (m *mockMemoryClient) Remember(context.Context, RememberArgs) (*RememberResult, error) {
	return nil, nil
}
func (m *mockMemoryClient) Search(context.Context, SearchArgs) (*SearchResult, error) {
	return nil, nil
}
func (m *mockMemoryClient) Recall(context.Context, RecallArgs) (*RecallResult, error) {
	return nil, nil
}
func (m *mockMemoryClient) Ask(context.Context, AskArgs, ProgressEmitter) (*AskResult, error) {
	return nil, nil
}
func (m *mockMemoryClient) IngestFile(ctx context.Context, args IngestFileArgs, progress ProgressEmitter) (*IngestResult, error) {
	if m.ingestFileFn != nil {
		return m.ingestFileFn(ctx, args, progress)
	}
	return &IngestResult{
		Status:     "completed",
		DocumentID: "doc-" + args.Path,
		Hash:       "hash-" + args.Path,
	}, nil
}
func (m *mockMemoryClient) IngestURL(context.Context, IngestURLArgs, ProgressEmitter) (*IngestURLResult, error) {
	return nil, nil
}
func (m *mockMemoryClient) ExtractAfterIngest(context.Context, ExtractAfterIngestArgs) (*ExtractAfterIngestResult, error) {
	return nil, nil
}
func (m *mockMemoryClient) Extract(context.Context, ExtractArgs, ProgressEmitter) (*ExtractResult, error) {
	return nil, nil
}
func (m *mockMemoryClient) Reflect(context.Context, ReflectArgs, ProgressEmitter) (*ReflectResult, error) {
	return nil, nil
}
func (m *mockMemoryClient) Consolidate(context.Context, ConsolidateArgs, ProgressEmitter) (*ConsolidateResult, error) {
	return nil, nil
}
func (m *mockMemoryClient) CreateBrain(context.Context, CreateBrainArgs) (*CreateBrainResult, error) {
	return nil, nil
}
func (m *mockMemoryClient) ListBrains(context.Context) (*ListBrainsResult, error) { return nil, nil }

func setupBatchTestServer(t *testing.T, client MemoryClient) *mcp.ClientSession {
	t.Helper()
	server := mcp.NewServer(&mcp.Implementation{Name: "test", Version: "0.0.0"}, nil)
	registerIngestBatch(server, client)

	serverTransport, clientTransport := mcp.NewInMemoryTransports()
	ctx, cancel := context.WithCancel(context.Background())
	t.Cleanup(cancel)

	if _, err := server.Connect(ctx, serverTransport, nil); err != nil {
		t.Fatalf("server connect: %v", err)
	}
	mcpClient := mcp.NewClient(&mcp.Implementation{Name: "test-client", Version: "0.0.1"}, nil)
	session, err := mcpClient.Connect(ctx, clientTransport, nil)
	if err != nil {
		t.Fatalf("client connect: %v", err)
	}
	t.Cleanup(func() { _ = session.Close() })
	return session
}

func callBatchTool(t *testing.T, session *mcp.ClientSession, args map[string]any) map[string]any {
	t.Helper()
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	res, err := session.CallTool(ctx, &mcp.CallToolParams{Name: "memory_ingest_batch", Arguments: args})
	if err != nil {
		t.Fatalf("CallTool: %v", err)
	}
	if res.IsError {
		txt := ""
		for _, c := range res.Content {
			if tc, ok := c.(*mcp.TextContent); ok {
				txt = tc.Text
			}
		}
		t.Fatalf("memory_ingest_batch returned error: %s", txt)
	}

	for _, c := range res.Content {
		if tc, ok := c.(*mcp.TextContent); ok {
			var out map[string]any
			if err := json.Unmarshal([]byte(tc.Text), &out); err != nil {
				t.Fatalf("unmarshal: %v", err)
			}
			return out
		}
	}
	t.Fatal("no text content in response")
	return nil
}

func TestIngestBatch_AllSucceed(t *testing.T) {
	client := &mockMemoryClient{}
	session := setupBatchTestServer(t, client)

	payload := callBatchTool(t, session, map[string]any{
		"files": []map[string]any{
			{"path": "/a.md"},
			{"path": "/b.md"},
			{"path": "/c.md"},
		},
	})

	if payload["total"] != float64(3) {
		t.Errorf("expected total=3, got %v", payload["total"])
	}
	if payload["succeeded"] != float64(3) {
		t.Errorf("expected succeeded=3, got %v", payload["succeeded"])
	}
	if payload["failed"] != float64(0) {
		t.Errorf("expected failed=0, got %v", payload["failed"])
	}
	results, ok := payload["results"].([]any)
	if !ok || len(results) != 3 {
		t.Fatalf("expected 3 results, got %v", payload["results"])
	}
	first, ok := results[0].(map[string]any)
	if !ok {
		t.Fatal("expected map result entry")
	}
	if first["status"] != "success" {
		t.Errorf("expected status=success, got %v", first["status"])
	}
}

func TestIngestBatch_PerFileErrorIsolation(t *testing.T) {
	client := &mockMemoryClient{
		ingestFileFn: func(_ context.Context, args IngestFileArgs, _ ProgressEmitter) (*IngestResult, error) {
			if args.Path == "/missing.md" {
				return nil, fmt.Errorf("file not found")
			}
			return &IngestResult{
				Status:     "completed",
				DocumentID: "doc-" + args.Path,
				Hash:       "hash-" + args.Path,
			}, nil
		},
	}
	session := setupBatchTestServer(t, client)

	payload := callBatchTool(t, session, map[string]any{
		"files": []map[string]any{
			{"path": "/a.md"},
			{"path": "/missing.md"},
			{"path": "/c.md"},
		},
	})

	if payload["succeeded"] != float64(2) {
		t.Errorf("expected succeeded=2, got %v", payload["succeeded"])
	}
	if payload["failed"] != float64(1) {
		t.Errorf("expected failed=1, got %v", payload["failed"])
	}

	results := payload["results"].([]any)
	second := results[1].(map[string]any)
	if second["status"] != "error" {
		t.Errorf("expected status=error for file 2, got %v", second["status"])
	}
	if second["error"] != "file not found" {
		t.Errorf("expected error message, got %v", second["error"])
	}
	third := results[2].(map[string]any)
	if third["status"] != "success" {
		t.Errorf("expected status=success for file 3, got %v", third["status"])
	}
}

func TestIngestBatch_DuplicateFiles(t *testing.T) {
	var mu sync.Mutex
	var callCount int
	client := &mockMemoryClient{
		ingestFileFn: func(_ context.Context, args IngestFileArgs, _ ProgressEmitter) (*IngestResult, error) {
			mu.Lock()
			callCount++
			n := callCount
			mu.Unlock()
			return &IngestResult{
				Status:     "completed",
				DocumentID: fmt.Sprintf("doc-%d", n),
				Hash:       fmt.Sprintf("hash-%d", n),
			}, nil
		},
	}
	session := setupBatchTestServer(t, client)

	payload := callBatchTool(t, session, map[string]any{
		"files": []map[string]any{
			{"path": "/same.md"},
			{"path": "/same.md"},
		},
	})

	if payload["total"] != float64(2) {
		t.Errorf("expected total=2, got %v", payload["total"])
	}
	if payload["succeeded"] != float64(2) {
		t.Errorf("expected succeeded=2, got %v", payload["succeeded"])
	}
	if payload["failed"] != float64(0) {
		t.Errorf("expected failed=0, got %v", payload["failed"])
	}
	mu.Lock()
	if callCount != 2 {
		t.Fatalf("expected 2 calls, got %d", callCount)
	}
	mu.Unlock()
}

func TestIngestBatch_BrainAppliedToAll(t *testing.T) {
	var mu sync.Mutex
	brainsSeen := []string{}
	client := &mockMemoryClient{
		ingestFileFn: func(_ context.Context, args IngestFileArgs, _ ProgressEmitter) (*IngestResult, error) {
			mu.Lock()
			brainsSeen = append(brainsSeen, args.Brain)
			mu.Unlock()
			return &IngestResult{Status: "completed", DocumentID: "d", Hash: "h"}, nil
		},
	}
	session := setupBatchTestServer(t, client)

	callBatchTool(t, session, map[string]any{
		"files": []map[string]any{
			{"path": "/a.md"},
			{"path": "/b.md"},
		},
		"brain": "test-brain",
	})

	mu.Lock()
	defer mu.Unlock()
	if len(brainsSeen) != 2 {
		t.Fatalf("expected 2 calls, got %d", len(brainsSeen))
	}
	for i, b := range brainsSeen {
		if b != "test-brain" {
			t.Errorf("call %d: expected brain=test-brain, got %s", i, b)
		}
	}
}
